libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
+        Node* left = UAVCAN_NULLPTR;
+        Node* right = UAVCAN_NULLPTR;
+        Node* equal_keys = UAVCAN_NULLPTR;
+    };
+
+    Node* root_;
+
+private:
+    size_t len_ ;
+
+    static int16_t heightOf(const Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return n->h;
+    }
+
+    void postOrderTraverseRecursively(Node* n, std::function<void(T*&)> forEach) {
+        if (n == UAVCAN_NULLPTR) {
+            return;
+        }
+
+        postOrderTraverseRecursively(n->left, forEach);
+        postOrderTraverseRecursively(n->right, forEach);
+        forEach(n->data);
+    }
+
+    Node* makeNode(T* payload) {
+        void* praw = this->allocator_.allocate(sizeof(Node));
+
+        if (praw == UAVCAN_NULLPTR) {
+            UAVCAN_TRACE("AvlTree", " OOM -- Can't allocate Node");
+            return UAVCAN_NULLPTR; // Push rejected
+        }
+
+        Node* node = new (praw) Node();
+        UAVCAN_ASSERT(node);
+
+        node->data = payload;
+        return node;
+    }
+
+    void deleteNode(Node*& n) {
+        if (n != UAVCAN_NULLPTR) {
+            n->~Node();
+            allocator_.deallocate(n);
+            n = UAVCAN_NULLPTR;
+        }
+    }
+
+    static int16_t balanceOf(Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return static_cast<int16_t>(heightOf(n->left) - heightOf(n->right));
+    }
+
+    static int16_t maxOf(int16_t a, int16_t b) {
```

Comment:
I don't think we need to write this ourselves: https://en.cppreference.com/w/cpp/algorithm/max

Reply:
If we care about semi-broken C++ implementations like that of NuttX (unless they fixed it since I last dealt with it), we must be extremely careful introducing new dependencies on STL. This is the reason why we have reinvented a huge wheel in [`uavcan/util/templates.hpp`](https://github.com/UAVCAN/libuavcan/blob/master/libuavcan/include/uavcan/util/templates.hpp) - none of this stuff one would expect to find in the standard library is available in NuttX. It is also why we can't easily just include `algorithm`, so I suggest keeping this code here unchanged.

I don't think that we should retain such drastic compatibility policies in bluesky, but this legacy code should remain compatible with older platforms.

Reply:
Okay. Shouldn't this go into templates.hpp then? I don't want to have multiple "min/max" functions lying around.

Reply:
> Shouldn't this go into templates.hpp then?

I guess it should.